// @rise4fun/docusaurus-plugin-application-insights/lib/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.validateOptions = void 0;
const tslib_1 = require("tslib");
const options_1 = require("./options");
const node_path_1 = require("node:path");
const validate_peer_dependencies_1 = tslib_1.__importDefault(require("validate-peer-dependencies"));
(0, validate_peer_dependencies_1.default)(__dirname);
function pluginApplicationInsights(_context, options) {
    const isProd = process.env.NODE_ENV === 'production';
    return {
        name: '@rise4fun/docusaurus-plugin-application-insights',
        getClientModules() {
            return isProd ? [(0, node_path_1.resolve)(__dirname, './analytics')] : [];
        },
        injectHtmlTags() {
            if (!isProd) {
                return {};
            }
            options = (0, options_1.normalizeConfig)(options);
            return {
                headTags: [
                    {
                        tagName: 'script',
                        innerHTML: `window.appInsightsPluginConfig = ${JSON.stringify(options)};`,
                    },
                ],
            };
        },
    };
}
exports.default = pluginApplicationInsights;
var options_2 = require("./options");
Object.defineProperty(exports, "validateOptions", { enumerable: true, get: function () { return options_2.validateOptions; } });
